tensorflow/core/common_runtime/simple_placer.cc

```diff
@@ -785,9 +785,23 @@ Status SimplePlacer::Run() {
 
     // Heuristic A application.
     if (IsGeneratorNode(node)) {
-      const Node* output = (*node->out_edges().begin())->dst();
+      EdgeSet::const_iterator it(node->out_edges().begin());
+
+      const Node* output = (*it)->dst();
       const string& output_device_name = output->assigned_device_name();
-      if (CanAssignToDevice(output_device_name, devices)) {
+      it++;
+
+      bool all_consumers_on_same_device(true);
+      while (it != node->out_edges().end()) {
+        if ((*it)->dst()->assigned_device_name() != output_device_name) {
+          all_consumers_on_same_device = false;
+          break;
+        }
+        it++;
+      }
+
```

Comment:
You might be able to replace this entire block of code with:

```
    bool same_device = std::all_of(node->edges().begin(), node->edges().end(), 
                                   [output_device_name](const Edge& e) { 
                                       return e->dst()->assigned_device_name() == output_device_name;
                                   }) 
```

If it turns out to be more complicated than this, let me know.